modules/yaml/src/main/scala/pureconfig/module/yaml/package.scala

```diff
@@ -0,0 +1,151 @@
+package pureconfig.module
+
+import java.io.{ IOException, Reader }
+import java.nio.file.{ Files, Path }
+
+import scala.collection.JavaConverters._
+import scala.reflect.ClassTag
+import scala.util.Try
+import scala.util.control.NonFatal
+
+import com.typesafe.config.{ ConfigValue, ConfigValueFactory }
+import org.yaml.snakeyaml.Yaml
+import org.yaml.snakeyaml.constructor.SafeConstructor
+import org.yaml.snakeyaml.error.{ Mark, MarkedYAMLException, YAMLException }
+import pureconfig.ConvertHelpers._
+import pureconfig._
+import pureconfig.error._
+import pureconfig.module.yaml.error.{ NonStringKeyFound, UnsupportedYamlType }
+
+package object yaml {
+
+  // Converts an object created by SnakeYAML to a Typesafe `ConfigValue`.
+  // (https://bitbucket.org/asomov/snakeyaml/wiki/Documentation#markdown-header-loading-yaml)
+  private[this] def yamlObjToConfigValue(obj: AnyRef): Either[ConfigReaderFailures, ConfigValue] = {
+
+    def aux(obj: AnyRef): Either[ConfigReaderFailures, AnyRef] = obj match {
+      case m: java.util.Map[Object @unchecked, Object @unchecked] =>
+        val entries = m.asScala.map {
+          case (k: String, v) => aux(v).right.map { v: AnyRef => k -> v }
+          case (k, _) => Left(ConfigReaderFailures(NonStringKeyFound(k.toString, k.getClass.getSimpleName)))
+        }
+        entries
+          .foldLeft(Right(Map.empty): Either[ConfigReaderFailures, Map[String, AnyRef]])(combineResults(_, _)(_ + _))
+          .right.map(_.asJava)
+
+      case xs: java.util.List[Object @unchecked] =>
+        xs.asScala.map(aux)
+          .foldRight(Right(Nil): Either[ConfigReaderFailures, List[AnyRef]])(combineResults(_, _)(_ :: _))
+          .right.map(_.asJava)
+
+      case s: java.util.Set[Object @unchecked] =>
+        s.asScala.map(aux)
+          .foldLeft(Right(Set.empty): Either[ConfigReaderFailures, Set[AnyRef]])(combineResults(_, _)(_ + _))
+          .right.map(_.asJava)
+
+      case _: java.lang.Integer | _: java.lang.Long | _: java.lang.Double | _: java.lang.String | _: java.lang.Boolean =>
+        Right(obj) // these types are supported directly by `ConfigValueFactory.fromAnyRef`
+
+      case i: java.math.BigInteger =>
+        Right(i.longValue(): java.lang.Long)
+
+      case _: java.util.Date | _: java.sql.Date | _: java.sql.Timestamp =>
+        Right(obj.toString)
+
+      case ba: Array[Byte] =>
+        Right(new String(ba))
```

Comment:
Hmm, good point... You are saying that there are some byte sequences `byteArr` such that `new String(byteArr).getBytes != byteArr`, is that right?

SnakeYAML reads YAML fields as byte arrays if a Base64 string is specified with the official `!!binary` (http://yaml.org/type/binary.html) annotation. Since PureConfig doesn't even have a default reader for byte arrays (which seems reasonable, as most config formats support only text), I say the best way to handle this is to provide readers a Base64 string, since that's the original content of the YAML file. What do you think?

Reply:
Yes, `new String(byteArr).getBytes != byteArr` is what I'm saying.

Converting to base 64 doesn't quite feel right to me. I can't think of why that would be more useful to the user than `Array[Byte]`. I suppose if the alternatives are no support or base 64, base 64 is preferable.

Would you like me to try to make the `AbstractConfigValue`-derived type wrapper approach I described above work?

Reply:
I don't think that having a specialized `AbstractConfigValue` makes much sense, given our current abstractions. Such a solution would require a user to make a `ConfigReader` that explicitly bypassed our `ConfigCursor` abstraction and checked if the `ConfigValue` was an instance of our new `ByteArrayConfigValue` in order to take advantage of YAML binary data. Such a `ConfigReader` would be compilation-dependent on this new module and would have specialized code for a backend, which seems the type of thing we want to avoid.

Choosing to represent such fields as Base64 would be a decision based not on usefulness, but on format consistency. Given a YAML file like this:

```yaml
gif_file: !!binary |
  R0lGODlhDAAMAIQAAP//9/X17unp5WZmZgAAAOfn515eXvPz7Y6OjuDg4J+fn5
  OTk6enp56enmlpaWNjY6Ojo4SEhP/++f/++f/++f/++f/++f/++f/++f/++f/+
  +f/++f/++f/++f/++f/++SH+Dk1hZGUgd2l0aCBHSU1QACwAAAAADAAMAAAFLC
  AgjoEwnuNAFOhpEMTRiggcz4BNJHrv/zCFcLiwMWYNG84BwwEeECcgggoBADs=
```

Assuming we don't support and don't want to support byte arrays in our AST, representing the data just as it is written in the file format (`R0lG`...`ADs=`) seems the most surprise-free behavior.

Reply:
(Personal opinion about this subject: even though I love the readability of the YAML format, I don't like many of their semantics; one of them is the notion of tags, that are expected to "tell" the libraries the concrete data types they should use. More often than not this leads to language-dependent config files and lack of portability, which is why I frequently advocate that parsers should concentrate on reading the data on config files just as it is represented - as very primitive data types - and let libraries such as PureConfig interpret them afterwards.)

Reply:
If we put the `AbstractConfigValue` and an implemented `ConfigReader[ByteArray]` for in the yaml project, then I think things would basically just work out.

I'm totally find boarding the base 64 train. I just wanted to talk out what the other side would look like.

Reply:
Besides my distaste for the design choice of extending the core AST to provide format-specific extensions, there's the practical aspects I can think of:
- `AbstractConfigValue`, as every class in `com.typesafe.config.impl`, is package-private. You would have to break package encapsulation and provide the new implementation in `com.typesafe.config.impl` too, which while feasible is rather hacky;
- What would happen if a user used the new `ConfigReader[Array[Byte]]` when loading a Typesafe config? What would happen if a user-provided `ConfigReader[Array[Byte]]` (for example, reading from Base64) was used with `loadYaml`?

My intention with this YAML module (and other backends) was always to support only a common subset of features, so that readers were always interchangeable. If we want to support such features, I'd prefer if we implemented a proper abstraction over the backends, with our own config AST and proper datatype extension support. But if you still think that we should investigate how an `AbstractConfigValue` implementation would look like, I can give it a try later :)

Reply:
Oh, I didn't notice that `AbstractConfigValue` is package private. Please forget I mentioned it.

Let's go base 64.

Reply:
@ruippeixotog, Just to make sure we're on the same page, I was going to wait to approve the PR until you'd made changes to encode the bytes as a base 64 string.

Reply:
Done in c0646a2. I also added some tests for null values in 6b8eaf9.